java/storage/connectors/redis/src/main/java/feast/storage/connectors/redis/retriever/OnlineRetriever.java

```diff
@@ -55,22 +59,47 @@ public OnlineRetriever(
   public List<List<Feature>> getOnlineFeatures(
       List<Map<String, ValueProto.Value>> entityRows,
       List<ServingAPIProto.FeatureReferenceV2> featureReferences,
-      List<String> entityNames) {
+      List<String> entityNames,
+      Map<String, List<String>> entityNamesPerFeatureView) {
+
+    List<String> featureViewNames =
+        featureReferences.stream()
+            .map(FeatureReferenceV2::getFeatureViewName)
+            .distinct()
+            .collect(Collectors.toList());
+
+    Set<List<String>> uniqueEntityKeyCombinations =
+        featureViewNames.stream().map(entityNamesPerFeatureView::get).collect(Collectors.toSet());
+
+    List<List<RedisProto.RedisKeyV2>> redisKeys =
+        entityRows.stream()
+            .map(
+                entityRow ->
+                    uniqueEntityKeyCombinations.stream()
+                        .map(
+                            entitiesToRetain -> {
+                              Map<String, Value> res = new HashMap<>(entityRow);
+                              res.keySet().retainAll(entitiesToRetain);
+                              return res;
+                            })
+                        .collect(Collectors.toList()))
+            .map(es -> RedisKeyGenerator.buildRedisKeys(this.project, es))
+            .collect(Collectors.toList());
 
-    List<RedisProto.RedisKeyV2> redisKeys =
-        RedisKeyGenerator.buildRedisKeys(this.project, entityRows);
     return getFeaturesFromRedis(redisKeys, featureReferences);
```

Comment:
we also need to group featureReferences by entities. Otherwise we'll be retrieving all features for all keys and that will cause redis latency degradation